bool port_initialize
bool invalid_msg
bool no_response